src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -95,19 +139,34 @@ public Http2Connection(Http2ConnectionContext context)
 
         public void OnInputOrOutputCompleted()
         {
-            _stopping = true;
+            if (_stateManager.Exchange(ConnectionState.Closed) != ConnectionState.Closed)
+            {
+                _frameWriter.WriteGoAwayAsync(_highestOpenedStreamId, Http2ErrorCode.NO_ERROR).ConfigureAwait(false).GetAwaiter().GetResult();
+            }
             _frameWriter.Complete();
         }
 
         public void Abort(ConnectionAbortedException ex)
         {
-            _stopping = true;
+            if (_stateManager.Exchange(ConnectionState.Closed) != ConnectionState.Closed)
+            {
+                _frameWriter.WriteGoAwayAsync(_highestOpenedStreamId, Http2ErrorCode.INTERNAL_ERROR).ConfigureAwait(false).GetAwaiter().GetResult();
```

Comment:
Alternatively, we need to convert everything to async, which means Http1Connection needs to change as well.

Reply:
I'm fine fire and forgetting the writing small frames. This isn't the only place we do it.

Nit: In Kestrel, we don't bother calling ConfigureAwait(false) before calling GetAwaiter().GetResult(). I don't see how it would do much since there's no continuation to post anyway.

Where we do use ConfigureAwait(false) is in the KestrelServer.StartAsync/StopAsync path because xunit's SynchronizationContext causes problems otherwise.